### PR TITLE
CIRC-1094 Remove pubsub unregistering

### PR DESCRIPTION
**Pull request by alexanderkurash**
[CIRC-1094](https://issues.folio.org/browse/CIRC-1094)

## Purpose
According to investigation in MODPUBSUB-147, when module is restarted it unregisters itself from pubsub after Okapi's call to `DELETE _tenant` and doesn't get registered again, because in some situations Okapi doesn't make a call to `POST _tenant` endpoint when module is started. This leads to mod-circulation being unable to publish events. mod-pubsub developers say that unregistering is unnecessary and some modules don't do it, because re-registering doesn't cause any issues. This was confirmed by re-registering manually on the snapshot env.

## Approach
Unregistering logic was removed from `DELETE _tenant`.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by adamdickmeiss** [comment]
Deploy and tenant init are two entirely different stages in the system. When starting a module (which may happen many times .. one for each instance).. You can assume nothing about okapi being present .. Only that environment is there.. Eg Kafka and Postgres.

Tenant init is something entirely different which is called - for each tenant - when a tenant is starting to *use* or *not use* your module.

**Comment by alexanderkurash** [comment]
> Deploy and tenant init are two entirely different stages in the system. When starting a module (which may happen many times .. one for each instance).. You can assume nothing about okapi being present .. Only that environment is there.. Eg Kafka and Postgres.
> 
> Tenant init is something entirely different which is called - for each tenant - when a tenant is starting to _use_ or _not use_ your module.

Thanks, Adam. Is it possible to create a situation when module is up and running (and is used by tenant), but the last tenant init (POST to _tenant) happened before the last DELETE to _tenant? It looks like this is what happened to mod-circualtion on snapshot on several occasions.

**Comment by adamdickmeiss** [comment]
> > Deploy and tenant init are two entirely different stages in the system. When starting a module (which may happen many times .. one for each instance).. You can assume nothing about okapi being present .. Only that environment is there.. Eg Kafka and Postgres.
> > Tenant init is something entirely different which is called - for each tenant - when a tenant is starting to _use_ or _not use_ your module.
> 
> Thanks, Adam. Is it possible to create a situation when module is up and running (and is used by tenant), but the last tenant init (POST to _tenant) happened before the last DELETE to _tenant? It looks like this is what happened to mod-circualtion on snapshot on several occasions.

I don't know what you mean

1. If a module is being used by a tenant, that should never happen prior to tenant init (create or update). Okapi should not route requests to a module before the tenant init succeeds.
2. is it possible that a module is called after tenant init (delete) ? Shouldn't happen.

I've tried to parse your comment but it doesn't make sense. Can you rephrase, please?

**Comment by alexanderkurash** [comment]
> I don't know what you mean
> 
> 1. If a module is being used by a tenant, that should never happen prior to tenant init (create or update). Okapi should not route requests to a module before the tenant init succeeds.
> 2. is it possible that a module is called after tenant init (delete) ? Shouldn't happen.
> 
> I've tried to parse your comment but it doesn't make sense. Can you rephrase, please?

I'll try. We register mod-circulation in pub-sub when module's `POST _tenant` is called and unregister it from pub-sub on `DELETE _tenant`. On multiple occasions we've seen on the snapshot env (according to Folijet team investigation) that mod-circulation is unregistered from pub-sub, though it is running and serving requests. So the only thing that was wrong with it was lack of pub-sub registration. Their guess is that module was restarted in such a way that `DELETE _tenant` was called (which caused the unregistering) but `POST _tenant` wasn't called afterwards (no registration). I'm trying to understand if this is possible and how. This PR is more of a workaround - getting rid of unregistering logic altogether. Pub-sub developers say that it's completely unnecessary and their modules never do it. But still, it would be nice to understand how this happened.

File: src/main/java/org/folio/circulation/resources/TenantActivationResource.java [hunk 1]
```diff
@@ -12,17 +14,15 @@
...
 
-import java.util.Map;
-
 public class TenantActivationResource {
 
   public void register(Router router) {
     RouteRegistration routeRegistration = new RouteRegistration("/_/tenant", router);
     routeRegistration.create(this::enableModuleForTenant);
-    routeRegistration.deleteAll(this::disableModuleForTenant);
+    routeRegistration.deleteAll(ctx -> noContent().writeTo(ctx.response()));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/TenantActivationResource.java:22]: Do we stiil have to implement `DELETE /_/tenant` if it does nothing?

File: src/test/java/api/TenantActivationResourceTests.java [hunk 2]
```diff
@@ -68,27 +67,12 @@ public void tenantActivationSucceedsWhenCanRegisterInPubSub() {
...
-    assertThat(getDeletedEventTypes().size(), is(6));
-    assertThat(getDeletedEventTypes(), hasItems(
-      EventTypeMatchers.ITEM_CHECKED_OUT,
-      EventTypeMatchers.ITEM_CHECKED_IN,
-      EventTypeMatchers.ITEM_DECLARED_LOST,
-      EventTypeMatchers.ITEM_CLAIMED_RETURNED,
-      EventTypeMatchers.LOAN_DUE_DATE_CHANGED,
-      EventTypeMatchers.LOG_RECORD
-    ));
+    assertThat(getDeletedEventTypes().size(), is(0));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/TenantActivationResourceTests.java:76]: Would it be better to assert that pubsub was not called? With PowerMock, for example.

**Comment by adamdickmeiss** [comment]
> > I don't know what you mean
> > 
> > 1. If a module is being used by a tenant, that should never happen prior to tenant init (create or update). Okapi should not route requests to a module before the tenant init succeeds.
> > 2. is it possible that a module is called after tenant init (delete) ? Shouldn't happen.
> > 
> > I've tried to parse your comment but it doesn't make sense. Can you rephrase, please?
> 
> I'll try. We register mod-circulation in pub-sub when module's `POST _tenant` is called and unregister it from pub-sub on `DELETE _tenant`. On multiple occasions we've seen on the snapshot env (according to Folijet team investigation) that mod-circulation is unregistered from pub-sub, though it is running and serving requests. So the only thing that was wrong with it was lack of pub-sub registration. Their guess is that module was restarted in such a way that `DELETE _tenant` was called (which caused the unregistering) but `POST _tenant` wasn't called afterwards (no registration). I'm trying to understand if this is possible and how. This PR is more of a workaround - getting rid of unregistering logic altogether. Pub-sub developers say that it's completely unnecessary and their modules never do it. But still, it would be nice to understand how this happened.

If you use Okapi to disable a module for a tenant (with or without purge), then okapi will insist on removing dependent modules as well. But mod-pubsub does not depend on mod-circulations interfaces. So as far as Okapi is concerned, it is safe to remove mod-circulation.

mod-pubsub should be prepared for a non-receiver.

If both mod-circulation (others) and mod-pubsub tampers with Kafka or a database at the same time, you need to be careful who controls and what does the work.

As an idea, when mod-circulation is disabled (or purged), it could ask mod-pubsub to do it. That's safe: mod-pubsub's interface is always available. So the mod-pubsub tampers with the data (kafka), but mod-circulation kindly asks it to stop. In that way, mod-pubsub could even push all remaining events out.. Before mod-circulation is really stopped.

(and no.. Do not make a circular dependency as a resolution)

File: src/main/java/org/folio/circulation/resources/TenantActivationResource.java (hunk 1 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/resources/TenantActivationResource.java:22]: In the [RMB 32 upgrade notes](https://github.com/folio-org/raml-module-builder/blob/master/doc/upgrading.md#version-320) they provide the whole tenant interface snippet and it does contain DELETE endpoint. Personally I don't know what will happen if we remove it and whether Okapi is ok with it's absence.

File: src/test/java/api/TenantActivationResourceTests.java (hunk 2 above)
Comment (alexanderkurash) [reply src/test/java/api/TenantActivationResourceTests.java:76]: PubSub is mocked by FakePubSub and we're checking here if we tried to delete any of our subscriptions/publications from this mocked service. Do you think it's substantially different?

File: src/main/java/org/folio/circulation/resources/TenantActivationResource.java (hunk 1 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/circulation/resources/TenantActivationResource.java:22]: As far as I remember, mod-circulation did not have any implementation of TenantAPI at all before it was integrated with mod-pubsub, and Okapi did not mind. But that's just me being curious, I guess we can keep this empty DELETE implementation.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/TenantActivationResource.java:22]: > As far as I remember, mod-circulation did not have any implementation of TenantAPI at all before it was integrated with mod-pubsub and Okapi did not mind.

It did not and Okapi was ok with that because it made no claim to. Now that it does, it needs to fulfil the full Tenant interface even if it does not need to do anything in some situations.

File: src/test/java/api/TenantActivationResourceTests.java (hunk 2 above)
Comment (OleksandrVidinieiev) [reply src/test/java/api/TenantActivationResourceTests.java:76]: I just looked at what FakePubsub does when we try to delete tenant, and you are right, it basically produces the same result I would have hoped to achieve with PowerMock.
